File: 2022/haskell/app/X4.hs
import Data.List.Split (splitOn)
import Data.Functor ((<&>))

main :: IO ()
main = do
  input :: [((Int, Int), (Int, Int))]
    <- readFile "../inputs/4"
    <&> lines
    <&> map (splitOn ",")
    <&> map (map (splitOn "-"))
    <&> map (\[[x, y], [a, b]] -> ((read x, read y), (read a, read b)))

  -- part one
  let contains (x, y) (a, b) = x <= a && y >= b

  print $ length [z | z@(x, y) <- input, contains x y || contains y x]

  -- part two
  print $ length [z | z@((x, y), (a, b)) <- input, any (`elem` [a..b]) [x..y]]

File: 2022/rust/src/x4.rs
use std::fs;

fn parse() -> Vec<((i32, i32), (i32, i32))> {
    // Giving up 😢
    // let input = fs::read_to_string("../inputs/4").expect("no file?");
    // let data = input.trim().split("\n")
    //            .map(|x| x.split(","))
    //            .map(|x| x.map(|y| y.split("-")))
    //            .map(|x| x.map(|y| y.map(|z| z.parse::<i32>())));

    let input = fs::read_to_string("../inputs/4").expect("no file?");
    let mut data = Vec::new();
    for line in input.trim().split("\n") {
        let mut tmp = Vec::new();
        for assigned in line.split(","){
            let range: Vec<&str> = assigned.split("-").collect();
            tmp.push((range[0].parse().unwrap(),
                      range[1].parse().unwrap()))
        }
        data.push((tmp[0], tmp[1]));
    }
    data
}

pub fn main() {
    let data = parse();

    // part one
    fn contains (((x, y), (a, b)): ((i32, i32), (i32, i32))) -> bool {
        (x <= a && y >=b) || (a <= x && b >= y)
    }

    println!("{:?}", data.iter().filter(|x| contains(**x)).collect::<Vec<_>>().len());

    // part two
    fn any_elem (((x, y), (a, b)): ((i32, i32), (i32, i32))) -> bool {
        (x..=y).any(|i| (a..=b).contains(&i))
    }

    println!("{:?}", data.iter().filter(|x| any_elem(**x)).collect::<Vec<_>>().len())

}
